akka-http-core/src/main/scala/akka/http/scaladsl/model/ws/WebSocketHandler.scala

```diff
@@ -0,0 +1,119 @@
+
+    def sinkAndSource(sink: Sink[T, Any], source: Source[T, Any]): SettingsBuilder
+    def source(source: Source[T, Any]): SettingsBuilder
+    def sink(sink: Sink[T, Any]): SettingsBuilder
+  }
+  trait SettingsBuilder {
+    def withMaxMessageSize(maxSize: Int): this.type
+    def withPerMessageTimeout(timeout: FiniteDuration): this.type
+    def withHalfClosedConnections(): this.type
+    def withIgnoreWrongMessageType(): this.type
+
+    def build(): Flow[Message, Message, Any]
+  }
+
+  // FIXME: make less arbitrary / get from config
+  val defaultSettings: Settings =
+    Settings(
+      maxMessageSize = 1000000,
+      perMessageTimeout = 10.seconds,
+      allowHalfClosed = false,
+      ignoreWrongMessageType = false
+    )
+  case class Settings(
+    maxMessageSize:         Int,
+    perMessageTimeout:      FiniteDuration,
+    allowHalfClosed:        Boolean,
+    ignoreWrongMessageType: Boolean
+  )
+  //private def builder[T](cons: T => Message, decons: Settings => PartialFunction[Message,Future[T]]): Builder[T] = ???
+  private def settingsBuilder[T](builder: Settings => Flow[Message, Message, Any]): SettingsBuilder =
+    SettingsBuilderImpl(defaultSettings, builder)
+
+  private case class SettingsBuilderImpl(settings: Settings, builder: Settings => Flow[Message, Message, Any]) extends SettingsBuilder {
+    override def withMaxMessageSize(maxSize: Int): this.type = set(_.copy(maxMessageSize = maxSize))
+    override def withPerMessageTimeout(timeout: FiniteDuration): this.type = set(_.copy(perMessageTimeout = timeout))
+    override def withHalfClosedConnections(): this.type = set(_.copy(allowHalfClosed = true))
+    override def withIgnoreWrongMessageType(): this.type = set(_.copy(ignoreWrongMessageType = true))
+
+    private def set(s: Settings => Settings): this.type = copy(settings = s(settings)).asInstanceOf[this.type]
+
+    override def build(): Flow[Message, Message, Any] = builder(settings)
+  }
+}
+
+object WebSocketHandlerExamples {
+  // simple text-based Request / Response pattern:
+  // synchronous
+
+  val strictHandler: String => String = txt => s"echo: $txt"
+  WebSocketHandler
+    .forText()
+    .mapMessages(strictHandler)
+    .build()
+
+  // would correspond to this flow:
+  implicit def ec: ExecutionContext = ???
+  implicit def mat: Materializer = ???
+  val handler: Flow[Message, Message, Any] =
+    Flow[Message]
+      .mapAsync[Seq[Message]](1) {
+        case tm: TextMessage =>
+          tm.toStrict(WebSocketHandler.defaultSettings.perMessageTimeout)
+            // FIXME: also check maxMessageSize here (or better in toStrict)
+            .map(t => TextMessage(strictHandler(t.text)) :: Nil)
+
+        // if ignoreWrongMessageType is false, which is currently the default
+        case _: BinaryMessage => Future.failed(new IllegalArgumentException("Unexpected binary WS message"))
```

Comment:
If would want to ignore BinaryMessages instead, it would have to drain the data stream here. A common un-pattern to use `.collect` leads to leaked Sources and temporarily stalled streams when messages are ignored.